spock-core/src/main/java/org/spockframework/runtime/extension/builtin/UnrollIterationNameProvider.java

```diff
@@ -24,58 +24,66 @@
 /**
  * @author Peter Niederwieser
  */
-public class UnrollNameProvider implements NameProvider<IterationInfo> {
+public class UnrollIterationNameProvider implements NameProvider<IterationInfo> {
   private static final Pattern EXPRESSION_PATTERN = Pattern.compile("#([a-zA-Z_$]([\\w$.]|\\(\\))*)");
 
   private final boolean assertUnrollExpressions = Boolean.getBoolean("spock.assertUnrollExpressions");
   private final FeatureInfo feature;
   private final Matcher expressionMatcher;
-  private int iterationCount;
 
-  public UnrollNameProvider(FeatureInfo feature, String namePattern) {
+  public UnrollIterationNameProvider(FeatureInfo feature, String namePattern) {
     this.feature = feature;
     expressionMatcher = EXPRESSION_PATTERN.matcher(namePattern);
   }
 
   // always returns a name
   @Override
   public String getName(IterationInfo iterationInfo) {
-    return nameFor(iterationInfo.getDataValues());
+    return nameFor(iterationInfo.getIterationIndex(), iterationInfo.getDataValues());
   }
 
-  String nameFor(Object... dataValues) {
+  String nameFor(int iterationIndex, Object... dataValues) {
     StringBuffer result = new StringBuffer();
     expressionMatcher.reset();
 
     while (expressionMatcher.find()) {
       String expr = expressionMatcher.group(1);
-      String value = evaluateExpression(expr, dataValues);
+      String value = evaluateExpression(expr, iterationIndex, dataValues);
       expressionMatcher.appendReplacement(result, Matcher.quoteReplacement(value));
     }
 
     expressionMatcher.appendTail(result);
-    iterationCount++;
     return result.toString();
   }
 
-  private String evaluateExpression(String expr, Object[] dataValues) {
+  private String evaluateExpression(String expr, int iterationIndex, Object[] dataValues) {
     String[] exprParts = expr.split("\\.");
     String firstPart = exprParts[0];
     Object result;
 
-    if ("featureName".equals(firstPart)) {
-      result = feature.getName();
-    } else if ("iterationCount".equals(firstPart)) {
-      result = String.valueOf(iterationCount);
-    } else {
-      int index = feature.getDataVariables().indexOf(firstPart);
-      if (index < 0) {
-        if (assertUnrollExpressions) {
-          throw new SpockAssertionError("Error in @Unroll, could not find matching variable for expression: " + expr);
+    switch (firstPart) {
+      case "featureName":
+        result = feature.getName();
+        break;
+
+      case "iterationCount":
+        //TODO: remove this case after some time, deprecated since Spock 2.0
+        System.err.println("WARNING: #iterationCount in unroll pattern is deprecated, use #iterationIndex instead");
```

Comment:
Thinking about it, the migration should be a simple `s/#iterationCount/iterationIndex`, so we don't really need it.
Spock has always been low on logging stuff.